### ISSUE TITLE
`as` prop

### ISSUE THREAD

**Issue by lluia** [zopaUK/react-components#24]
Many component libraries implement the now famous `as` prop:
- [Reakit `as`](https://reakit.io/docs/composition/#as-prop)
- [RBX `as`](https://dfee.github.io/rbx/base/generic#core-properties)
- [Material UI `<Box component />`](https://material-ui.com/components/box/#overriding-material-ui-components)
- [Grommet `<Box as />`](https://v2.grommet.io/box#as)
- [Semantic UI `as`](https://react.semantic-ui.com/augmentation)

It allows you to control declaratively what **HTML root tag** gets rendered by a component.

The `as` prop would simplify the API of this library by removing the need for the existence of some components we have now:
```jsx
/**
 * I believe there would not be a need for <InputLabel /> anymore
 */
<InputLabel >
<Text as="label" />

/**
 * We could export <Link /> as a convenience but only mapping to `<Text />`
 */
<Link /> 
<Text as="a" />


/**
 * We could still export a <Heading size={'l' | 'm' | 's'} /> component as a convenience but internally it 
 * would just map to `<Text />`
 */
<Header1 /> <Header2 /> <Header3 />
<Text as="h1" variant="display" /> 
<Text as="h2" variant="display" />
<Text as="h3" variant="display" />

/**
 * I think there wouldn't be a need for this ones?
 */
<Title1 /> <Title2 /> <Title3 />
<Text as="h1" /> <Text as="h2" /> <Text as="h3" />
```

**Comment by thegrinder** [comment]
@lluia if we spread props then we get this out of the box with styled-components https://www.styled-components.com/docs/api#as-polymorphic-prop

**Comment by lluia** [comment]
@thegrinder nice and good catch! 🎊 ❤️  

@poteirard are you happy if we work on this? I guess we'll only need to make sure that:
- `render` root node is **always** a styled component
- props are always spread on this ☝️ node

**Comment by poteirard** [comment]
Maybe it makes more sense to have only `<Text />` component and the prop `as` as required in it? Also maybe after the rebranding the typography components doesn't make sense (I don't know)

**Comment by thegrinder** [comment]
`span` seems like a sensible default to me to be honest. There would be only a few cases where you actually wanna use `<p>` tag. With the rebranding, we'd still need it but would just change the sizes accordingly

**Comment by lluia** [comment]
I can talk to the design team when I'm in London next week, and explain our idea to simplify our typography through `<Text />` ( _they too wanted to simplify last time I checked with them_ ) and maybe get some spec ready 🧐 📐

**Comment by lluia** [comment]
Closing as @thegrinder mentioned, this is already provided by [`styled-components`](https://www.styled-components.com/docs/api#as-polymorphic-prop).

**Comment by poteirard** [comment]
Reopening since TS complains about types related to `as`.

At the moment we can use `as` the problem is that the types do not interfere properly. 

For example in my case I want to have:

```ts
<Navbar.Link as={ReactRouterLink} to={some_page_url}>

**Comment by lluia** [comment]
Addressed on [`2.0`](https://github.com/zopaUK/react-components/pull/54) ☝️
